src/core/sphinx/crypto.rs

```diff
@@ -18,146 +18,230 @@
-	pub header_encryption: [u8; STREAM_KEY_SIZE],
-	pub header_encryption_iv: [u8; STREAM_IV_SIZE],
-	pub payload_encryption: [u8; SPRP_KEY_SIZE],
-	pub blinding_factor: [u8; KEY_SIZE],
-}
+	pub fn mac_key(&self) -> &MacKey {
+		self.split().0
+	}
 
-unsafe impl bytemuck::Zeroable for PacketKeys {}
+	pub fn header_encryption_key(&self) -> &HeaderEncryptionKey {
+		self.split().1
+	}
 
-unsafe impl bytemuck::Pod for PacketKeys {}
+	pub fn delay_seed(&self) -> &DelaySeed {
+		self.split().2
+	}
+}
 
-unsafe impl bytemuck::Zeroable for KdfOutput {}
+pub const PAYLOAD_ENCRYPTION_KEY_SIZE: usize = 192;
+pub type PayloadEncryptionKey = [u8; PAYLOAD_ENCRYPTION_KEY_SIZE];
 
-unsafe impl bytemuck::Pod for KdfOutput {}
+pub fn derive_payload_encryption_key(kx_shared_secret: &KxSharedSecret) -> PayloadEncryptionKey {
+	let mut derived = [0; PAYLOAD_ENCRYPTION_KEY_SIZE];
+	derive_secret(&mut derived, kx_shared_secret, PAYLOAD_ENCRYPTION_KEY_PERSONA);
+	derived
+}
 
-#[derive(Copy, Clone)]
-#[repr(transparent)]
-struct KdfOutput([u8; KDF_OUTPUT_SIZE]);
+////////////////////////////////////////////////////////////////////////////////
+// MAC computation
+////////////////////////////////////////////////////////////////////////////////
 
-/// `kdf` takes the input key material and returns the Sphinx Packet keys.
-pub fn kdf(input: &RawKey) -> PacketKeys {
-	let output = hkdf_expand(input, String::from(KDF_INFO_STR).into_bytes().as_slice());
-	bytemuck::cast(KdfOutput(output))
+pub fn compute_mac(encrypted_header: &[u8], pad: &[u8], key: &MacKey) -> Mac {
+	let mut h = Blake2bMac::<U16>::new_from_slice(key).expect("Key size is fixed and small enough");
+	h.update(encrypted_header);
+	h.update(pad);
+	h.finalize().into_bytes().into()
 }
 
-pub fn hkdf_expand(prk: &[u8], info: &[u8]) -> [u8; KDF_OUTPUT_SIZE] {
-	let mut output = [0u8; KDF_OUTPUT_SIZE];
-	let hk = Hkdf::<Sha256>::from_prk(prk).unwrap();
-	hk.expand(info, &mut output).unwrap();
-	output
+pub fn mac_ok(mac: &Mac, encrypted_header: &EncryptedHeader, key: &MacKey) -> bool {
+	let mut h = Blake2bMac::<U16>::new_from_slice(key).expect("Key size is fixed and small enough");
+	h.update(encrypted_header);
+	h.verify(mac.into()).is_ok()
 }
 
-/// Hash the input.
-pub fn hash(input: &[u8; KEY_SIZE]) -> [u8; HASH_OUTPUT_SIZE] {
-	use blake2::digest::{FixedOutput, Update};
-	type Blake2b256 = blake2::Blake2b<blake2::digest::typenum::U32>;
-	let mut r = [0u8; HASH_OUTPUT_SIZE];
-	let mut ctx = Blake2b256::default();
-	ctx.update(input);
-	let hash = ctx.finalize_fixed();
-	r.copy_from_slice(&hash);
-	r
+////////////////////////////////////////////////////////////////////////////////
+// Header encryption
+////////////////////////////////////////////////////////////////////////////////
+
+pub fn apply_header_encryption_keystream(data: &mut [u8], key: &HeaderEncryptionKey) {
+	// Key is only used once, so fine for nonce to be 0
+	let mut c = ChaCha20::new(key.into(), &[0; 8].into());
```

Comment:
Yeah I switched to ChaCha mainly because it's what `lioness` uses, and it's simpler than AES in SW (so seems like a good choice for eg Wasm). To be fair it is actually slower than AES if your HW has AES instructions, though I don't really think speed matters that much here -- the elliptic curve stuff dominates.